agent/app/agent_capability.go

```diff
@@ -316,6 +345,92 @@ func (agent *ecsAgent) appendTaskENICapabilities(capabilities []*ecs.Attribute)
 	return capabilities
 }
 
+func (agent *ecsAgent) appendExecCapabilities(capabilities []*ecs.Attribute) ([]*ecs.Attribute, error) {
+	// for an instance to be exec-enabled, it needs resources needed by SSM (binaries, configuration files and certs)
+	// the following bind mounts are defined in ecs-init and added to the ecs-agent container
+
+	capabilityExecRootDir := filepath.Join(capabilityDepsRootDir, capabilityExec)
+	binDir := filepath.Join(capabilityExecRootDir, capabilityExecBinRelativePath)
+	configDir := filepath.Join(capabilityExecRootDir, capabilityExecConfigRelativePath)
+	certsDir := filepath.Join(capabilityExecRootDir, capabilityExecCertsRelativePath)
+
+	// top-level folders, /bin, /config, /certs
+	dependencies := map[string][]string{
+		binDir:    []string{},
+		configDir: []string{},
+		certsDir:  capabilityExecRequiredCerts,
+	}
+	if exists, err := dependenciesExist(dependencies); err != nil || !exists {
+		return capabilities, err
+	}
+
+	// ssm binaries are stored in /bin/<version>/, 1 version is downloaded by ami builder for ECS instances
+	binDependencies := map[string][]string{}
+	// child folders named by version inside binDir, e.g. 3.0.236.0
+	binFolders, err := getSubDirectories(binDir)
+	if err != nil {
+		return capabilities, err
+	}
+	for _, binFolder := range binFolders {
+		// check for the same set of binaries for all versions for now
+		// change this if future versions of ssm agent require different binaries
+		versionSubDirectory := filepath.Join(binDir, binFolder)
+		binDependencies[versionSubDirectory] = capabilityExecRequiredBinaries
+	}
+	if len(binDependencies) < 1 {
+		return capabilities, nil
+	}
+	if exists, err := dependenciesExist(binDependencies); err != nil || !exists {
+		return capabilities, err
+	}
+
+	return appendNameOnlyAttribute(capabilities, attributePrefix+capabilityExec), nil
+}
+
+func defaultGetSubDirectories(path string) ([]string, error) {
+	var subDirectories []string
+
+	fileInfos, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+	for _, fileInfo := range fileInfos {
+		if fileInfo.IsDir() {
```

Comment:
You have to exclude directories that don't conform to valid ssm versions here. We can't assume that our dependencies dir will always be pristine, someone could create files inside the dir and we can't error out because of that.

Reply:
good point, I'll add a quick check here for folders matching this regex: `\d+(\.\d+)*`, but at some point I think it'd be beneficial to reuse code that we're currently adding to `agent/engine/execcmd/agent_version.go`

Reply:
I think regex needs to be `^\d+(\.\d+)*$` 
https://regex101.com/r/yi3sgO/1

Reply:
thanks for pointing it out! add anchor tokens